Tensile/KernelWriterAssembly.py

```diff
@@ -7684,15 +7696,29 @@ def globalReadGuardK(self, kernel, tP, vregSetIdx):
                   offsetVgpr = self.guardZeroPad(kernel, tP, codeMod, offsetVgpr, soffset, addrV, perp, sPerp, para, sPara)
                   kStr += str(codeMod)
 
-                unrollMirrorWithSoffset = kernel["ProblemType"]["IndicesSummation"][self.unrollIdx] in problemType["MirrorDims%s"%tc] and soffset != "0"
-                # ScalarGlobalReadOffset should be negative value with unroll mirroring.
-                # However, buffer_load uses soffset as uint value, so GRO - SGRO, SGRO = 0
-                if unrollMirrorWithSoffset:
+                if mirrorDim:
                   codeMod = Code.Module("mirrorIdx%u"%loopCnt)
-                  codeMod.addInst("_v_sub_u32", vgpr(offsetVgpr), vgpr(offsetVgpr), soffset, "mirror unroll: GRO=GRO-SGRO, soffset=0")
+                  codeMod.addInst("v_mov_b32", vgpr(mirrorOffsetTemp), vgpr(offsetVgpr), "copy offset to clamp mirror dim")
+                  offsetVgpr = mirrorOffsetTemp
+
+                  # ScalarGlobalReadOffset should be negative value with unroll mirroring.
+                  # However, buffer_load uses soffset as uint value, so GRO - SGRO, SGRO = 0
+                  if soffset != "0":
+                    codeMod.addInst("_v_sub_u32", vgpr(offsetVgpr), vgpr(offsetVgpr), soffset, "mirror unroll: GRO=GRO-SGRO, soffset=0")
+                    # soffset_prev = soffset
+                    soffset = "0"
+
+                  # check for pre-padding
+                  prePad = 0
+                  if self.groOffsetInMacroTile:
+                    prePad = self.srdShiftLeft[tc]
+                    if tc == "A":
+                      prePad *= self.tPA["bpe"]
+                    elif tc == "B":
+                      prePad *= self.tPB["bpe"]
+
+                  codeMod.addInst("v_max_i32", vgpr(offsetVgpr), prePad, vgpr(offsetVgpr), "clamp to beginning of buffer")
```

Comment:
I think it is better to stick to unsigned calculation.
It is technically possible to have >=2^31 values in vgpr(offsetVgpr) or soffset.
(if lda is very large)
We can do like this
   tmp = offsetVgpr - 4 # remove prepad
   tmp = max_u32(tmp, soffset)
   tmp = tmp - soffset
   tmp = tmp + 4

Reply:
I don't think mirror dims feature ever worked for offset > 2^31 because sgpr offset was always signed i32. Looking through the code, there's a lot of other code that would also have to change to get it to work for u32. I might file a feature ticket to do that work separately from this change.

Reply:
> I don't think mirror dims feature ever worked for offset > 2^31 because sgpr offset was always signed i32. Looking through the code, there's a lot of other code that would also have to change to get it to work for u32. I might file a feature ticket to do that work separately from this change.

How about vgprGlobalReadOffset?
Even if sgpr offset is within 2^31, v_max_i32 can be wrong if vgprGlobalReadOffset is already >2^31.